[UR][L0] Add L0 teardown notification support and simply teardown with proxy loader support

nrspruit opened the pull request:
- Add support for L0 teardown notification in the static L0 loader.
- Add loading of the L0 dynamic loader in the windows proxy loader to ensure that the L0 dynamic loader is usable during teardown.

nrspruit commented:
related to https://github.com/oneapi-src/level-zero/pull/333

cperkinsintel commented:
@nrspruit  - the test failures reported here seem legit.    But they are strange.

The `buffer_create.cpp` test works fine for me if I build and run it manually. Everything working exactly as it should. But when run from llvm-lit, I can get this output:

```# | 1 L0 Drivers found.
# | 
# | ZE ---> zeDeviceGet(ZeDrivers[I], &ZeDeviceCount, nullptr)
# | ZE ---> zeDeviceGet(ZeDrivers[I], &ZeDeviceCount, ZeDevices.data())
# | ZE ---> zeDeviceGetProperties(ZeDevices[D], &device_properties)
# | ZE ---> zeDriverGetApiVersion(ZeDriver, &ZeApiVersion)
# | ZE ---> zeDriverGetExtensionProperties(ZeDriver, &Count, nullptr)
# | ZE ---> zeDriverGetExtensionProperties(ZeDriver, &Count, ZeExtensions.data())
# | ZE ---> zelLoaderTranslateHandle(ZEL_HANDLE_DRIVE
# | ...
# `---data was truncated--------
```

The `data was truncated` comes from llvm-lit.  Essentially, there seemed to have been an error when `zelLoaderTranslateHandle` was called. I think the next argument it is supposed to log is the "ZeDriver", so there might be something going on.

nrspruit commented:
> @nrspruit - the test failures reported here seem legit. But they are strange.
> 
> The `buffer_create.cpp` test works fine for me if I build and run it manually. Everything working exactly as it should. But when run from llvm-lit, I can get this output:
> 
> ```
> # | 
> # | ZE ---> zeDeviceGet(ZeDrivers[I], &ZeDeviceCount, nullptr)
> # | ZE ---> zeDeviceGet(ZeDrivers[I], &ZeDeviceCount, ZeDevices.data())
> # | ZE ---> zeDeviceGetProperties(ZeDevices[D], &device_properties)
> # | ZE ---> zeDriverGetApiVersion(ZeDriver, &ZeApiVersion)
> # | ZE ---> zeDriverGetExtensionProperties(ZeDriver, &Count, nullptr)
> # | ZE ---> zeDriverGetExtensionProperties(ZeDriver, &Count, ZeExtensions.data())
> # | ZE ---> zelLoaderTranslateHandle(ZEL_HANDLE_DRIVE
> # | ...
> # `---data was truncated--------
> ```
> 
> The `data was truncated` comes from llvm-lit. Essentially, there seemed to have been an error when `zelLoaderTranslateHandle` was called. I think the next argument it is supposed to log is the "ZeDriver", so there might be something going on. It didn't even output the last "R"
> 
> The test failed on Arc, I think. But I can repro the failure on pvc-01.

that error has existed since before this change, so it may be a real issue, but it is not part of this PR.

See here:
https://github.com/intel/llvm/actions/runs/15042177600/job/42277602962

cperkinsintel commented:
Test failures are known: https://github.com/intel/llvm/issues/18463

gmlueck commented:
I assume this PR has some sort of merge problem?  Seems like a lot of unrelated things changed.

nrspruit commented:
> I assume this PR has some sort of merge problem? Seems like a lot of unrelated things changed.

the rebase failed, let me fix

nrspruit commented:
> > I assume this PR has some sort of merge problem? Seems like a lot of unrelated things changed.
> 
> the rebase failed, let me fix

fixed, for some reason the pulldown was included with I rebased.

jchlanda commented on sycl/ur_win_proxy_loader/ur_win_proxy_loader.cpp:
```diff
@@ -168,6 +168,11 @@ void preloadLibraries() {
   loadAdapter(UR_LIBRARY_NAME(adapter_native_cpu));
+  // Load the Level Zero loader dynamic library to ensure it is loaded during
+  // the runtime. This is necessary to avoid the level zero loader from being
+  // unloaded prematurely. the Only trusted loader is the one that is loaded
```
```suggestion
  // unloaded prematurely. The only trusted loader is the one that is loaded
```

maarquitos14 reviewed (approved):
LGTM.

nrspruit commented:
@intel/llvm-gatekeepers , please merge when available, this enables usage of an improved short term and long term solution to the teardown issues in SYCL when using L0.
